### PR TITLE
Node replacement - Reactive(IPI)

### PR DESCRIPTION
**Pull request by PrasadDesala**
Signed-off-by: Prasad Desala <tdesala@redhat.com>

### CODE REVIEW

File: ocs_ci/ocs/node.py [hunk 1]
```diff
@@ -359,3 +362,77 @@ def get_node_resource_utilization_from_oc_describe(nodename=None, node_type='wor
         }
 
     return utilization_dict
+
+
+def get_osd_running_nodes():
+    """
+    Gets the osd running node names
```
Comment (ebenahar) [inline ocs_ci/ocs/node.py]: Add an empty line below

File: ocs_ci/ocs/node.py [hunk 2]
```diff
@@ -359,3 +362,77 @@ def get_node_resource_utilization_from_oc_describe(nodename=None, node_type='wor
...
+    Returns:
+        list: OSD node names
```
Comment (ebenahar) [inline ocs_ci/ocs/node.py]: add an empty line below

File: ocs_ci/ocs/node.py [hunk 3]
```diff
@@ -359,3 +362,77 @@ def get_node_resource_utilization_from_oc_describe(nodename=None, node_type='wor
...
+    """
+    osd_pods_obj = pod.get_osd_pods()
```
Comment (ebenahar) [inline ocs_ci/ocs/node.py]: you can replace lines 373-377 with:
```python
return [pod.get_pod_node(osd_node).name for osd_node in pod.get_osd_pods()]
```

File: ocs_ci/ocs/node.py [hunk 4]
```diff
@@ -359,3 +362,77 @@ def get_node_resource_utilization_from_oc_describe(nodename=None, node_type='wor
...
+
+def get_app_pod_running_nodes(pod_obj):
+    """
+    Gets the app pod running node names
+    Args:
+        pod_obj (list): List of app pod objects
+    Returns:
+        list: App pod running node names
+    """
+    app_pod_node_name = []
```
Comment (ebenahar) [inline ocs_ci/ocs/node.py]: you can replace lines 388-3391 with:
```python
return [pod.get_pod_node(node).name for node in pod_obj]
```

File: ocs_ci/ocs/node.py [hunk 5]
```diff
@@ -359,3 +362,77 @@ def get_node_resource_utilization_from_oc_describe(nodename=None, node_type='wor
...
+    osd_pods_obj = pod.get_osd_pods()
+    osd_node_name = []
+    for osd_node in osd_pods_obj:
+        osd_node_name.append(pod.get_pod_node(osd_node).name)
+    return osd_node_name
+
+
+def get_app_pod_running_nodes(pod_obj):
+    """
+    Gets the app pod running node names
```
Comment (ebenahar) [inline ocs_ci/ocs/node.py]: add an empty line below

File: ocs_ci/ocs/node.py [hunk 6]
```diff
@@ -359,3 +362,77 @@ def get_node_resource_utilization_from_oc_describe(nodename=None, node_type='wor
...
+    Args:
+        pod_obj (list): List of app pod objects
```
Comment (ebenahar) [inline ocs_ci/ocs/node.py]: add an empty line below

File: ocs_ci/ocs/node.py [hunk 7]
```diff
@@ -359,3 +362,77 @@ def get_node_resource_utilization_from_oc_describe(nodename=None, node_type='wor
...
+    Returns:
+        list: App pod running node names
```
Comment (ebenahar) [inline ocs_ci/ocs/node.py]: add an empty line below

File: ocs_ci/ocs/node.py [hunk 8]
```diff
@@ -359,3 +362,77 @@ def get_node_resource_utilization_from_oc_describe(nodename=None, node_type='wor
...
+    """
+    app_pod_node_name = []
+    for node in pod_obj:
+        app_pod_node_name.append(pod.get_pod_node(node).name)
+    return app_pod_node_name
+
+
+def get_both_osd_and_app_pod_running_node(
+        osd_running_nodes, app_pod_running_nodes
```
Comment (ebenahar) [inline ocs_ci/ocs/node.py]: delete extra indentation

File: ocs_ci/ocs/node.py [hunk 9]
```diff
@@ -359,3 +362,77 @@ def get_node_resource_utilization_from_oc_describe(nodename=None, node_type='wor
...
+):
+    """
```
Comment (ebenahar) [inline ocs_ci/ocs/node.py]: add empty lines, as I commented above

File: ocs_ci/ocs/node.py [hunk 10]
```diff
@@ -359,3 +362,77 @@ def get_node_resource_utilization_from_oc_describe(nodename=None, node_type='wor
...
+     Returns:
+         list: Both OSD and app pod running node names
+     """
+    common_nodes = list(set(osd_running_nodes) & set(app_pod_running_nodes))
+    log.info(f"Common node is {common_nodes}")
+    return common_nodes
+
+
+def node_network_failure(node_names, wait=True):
+    """
```
Comment (ebenahar) [inline ocs_ci/ocs/node.py]: add empty lines, as I commented above

File: ocs_ci/ocs/resources/pod.py [hunk 11]
```diff
@@ -1176,3 +1176,19 @@ def get_noobaa_pods(noobaa_label=constants.NOOBAA_APP_LABEL, namespace=None):
     noobaa_pods = [Pod(**noobaa) for noobaa in noobaas]
 
     return noobaa_pods
+
+
+def wait_for_dc_app_pods_to_reach_running_state(dc_pod_obj):
+    """
```
Comment (ebenahar) [inline ocs_ci/ocs/resources/pod.py]: add empty lines, as I commented above

**Comment by PrasadDesala** [comment]
> PR validation [Job UNSTABLE](https://ocs4-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/qe-trigger-test-pr/1224//testReport) (some or all tests failed).

This PR validation seems to be failing because of https://bugzilla.redhat.com/show_bug.cgi?id=1810014

File: ocs_ci/ocs/node.py (hunk 1 above)
Comment (PrasadDesala) [reply ocs_ci/ocs/node.py]: Done

File: ocs_ci/ocs/node.py (hunk 2 above)
Comment (PrasadDesala) [reply ocs_ci/ocs/node.py]: Done

File: ocs_ci/ocs/node.py (hunk 5 above)
Comment (PrasadDesala) [reply ocs_ci/ocs/node.py]: Done

File: ocs_ci/ocs/node.py (hunk 6 above)
Comment (PrasadDesala) [reply ocs_ci/ocs/node.py]: Done

File: ocs_ci/ocs/node.py (hunk 7 above)
Comment (PrasadDesala) [reply ocs_ci/ocs/node.py]: Done

File: ocs_ci/ocs/node.py (hunk 9 above)
Comment (PrasadDesala) [reply ocs_ci/ocs/node.py]: Done

File: ocs_ci/ocs/node.py (hunk 10 above)
Comment (PrasadDesala) [reply ocs_ci/ocs/node.py]: Done

File: ocs_ci/ocs/resources/pod.py (hunk 11 above)
Comment (PrasadDesala) [reply ocs_ci/ocs/resources/pod.py]: Done

**Comment by ebenahar** [comment]
@PrasadDesala please rebase

**Comment by PrasadDesala** [comment]
> @PrasadDesala please rebase

Done

File: tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+        # Get the failure node obj
+        failure_node_obj = get_node_objs(node_names=[common_nodes[0]])
+
+        # Induce failure on the selected failure node
+        log.info(f"Inducing failure on node {failure_node_obj[0].name}")
+        if failure == "power off":
+            # Power off AWS worker node instance
+            nodes.stop_nodes(failure_node_obj, wait=True)
+            log.info(f"Successfully powered off node: {failure_node_obj[0].name}")
+        elif failure == "network failure":
```
Comment (fbalak) [inline tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: Can there be more failures? Would another `else` branch make sense?

Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_reactive_aws_ipi.py]: Yes, Network failure is one of the way to make node into "Not Ready". There can be other ways to bring it down.

**Comment by vasukulkarni** [comment]
Nice!

**Pull request merged**